### PR TITLE
CIRC-2072 Create a facade for instance search

### PR DESCRIPTION
**Pull request by MagzhanArtykov**
Covers: [CIRC-2072](https://folio-org.atlassian.net/browse/CIRC-2072)

## Purpose
 Create a facade for instance search

## Approach
1. Create a new endpoint that would accept instanceId: /circulation/items-by-instance/{id}
2. Make a call to mod-search to get info about this instance and all of its items
3. Make calls to mod-inventory-storage to get item info
4. Combine everything into a single response.

### CODE REVIEW

File: src/main/java/org/folio/circulation/domain/InstanceExtended.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+@ToString(onlyExplicitlyIncluded = true)
+public class InstanceExtended {
+
+  private static final Logger log = LogManager.getLogger(MethodHandles.lookup().lookupClass());
+
+  JsonObject representation;
+
+  String id;
+
+  @NonNull Collection<Item> items;
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/domain/InstanceExtended.java]: Redundant blank lines

File: src/main/java/org/folio/circulation/domain/Item.java [hunk 2]
```diff
@@ -415,4 +415,15 @@ public String getLendingLibraryCode() {
...
   }
+
+  public String getTenantId() {
+    return getProperty(itemRepresentation, "tenantId");
+  }
+
+  public Item changeTenantId(String tenantId) {
+    if(tenantId != null) {
+      write(itemRepresentation, "tenantId", tenantId);
+    }
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/Item.java:426]: no need to check for null, `write()` already does that

File: src/main/java/org/folio/circulation/domain/SearchInstance.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import org.apache.logging.log4j.Logger;
+import org.folio.circulation.domain.representations.ItemSummaryRepresentation;
+import org.folio.circulation.storage.mappers.ItemMapper;
+
+import java.lang.invoke.MethodHandles;
+import java.util.Collection;
+import java.util.List;
+
+import static org.folio.circulation.support.json.JsonObjectArrayPropertyFetcher.mapToList;
+import static org.folio.circulation.support.json.JsonPropertyWriter.write;
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/SearchInstance.java:18]: Wrong import order

File: src/main/java/org/folio/circulation/domain/SearchInstance.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+@ToString(onlyExplicitlyIncluded = true)
+public class SearchInstance {
+
+  private static final Logger log = LogManager.getLogger(MethodHandles.lookup().lookupClass());
+
+  JsonObject representation;
+
+  String id;
+
+  @NonNull Collection<Item> items;
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/SearchInstance.java:30]:
```suggestion
  JsonObject representation;
  String id;
  @NonNull Collection<Item> items;
```

File: src/main/java/org/folio/circulation/infrastructure/storage/SearchRepository.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+@AllArgsConstructor
+public class SearchRepository {
+
+  private static final Logger log = LogManager.getLogger(MethodHandles.lookup().lookupClass());
+
+  private final ItemRepository itemRepository;
+  private final CollectionResourceClient searchClient;
+
+
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/infrastructure/storage/SearchRepository.java:31]: Redundant empty line

File: src/main/java/org/folio/circulation/infrastructure/storage/SearchRepository.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  public SearchRepository(Clients clients) {
+    this(new ItemRepository(clients), clients.searchClient());
+  }
+
+  public CompletableFuture<Result<SearchInstance>> getInstanceWithItems(String query) {
+    log.debug("getInstanceWithItems:: query {}", query);
+    return searchClient.getManyWithQueryStringParameters(Map.of("expandAll",
+        "true", "query", URLEncoder.encode(query,
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/infrastructure/storage/SearchRepository.java:39]: There is Folio's `StringUtil#encodeUrl` for that

File: src/main/java/org/folio/circulation/infrastructure/storage/SearchRepository.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+      .thenApply(mapResult(MultipleRecords::firstOrNull))
+      .thenCompose(r -> r.after(this::updateItemDetails));
+  }
+
+  private Result<MultipleRecords<SearchInstance>> mapResponseToInstances(Response response) {
+    return MultipleRecords.from(response, SearchInstance::from, "instances");
+  }
+
+  private CompletableFuture<Result<SearchInstance>> updateItemDetails(SearchInstance searchInstance) {
+    log.debug("updateItemDetails:: searchInstance {}", searchInstance);
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/infrastructure/storage/SearchRepository.java:51]:
```suggestion
    log.debug("updateItemDetails:: searchInstance {}", () -> searchInstance);
```

File: src/main/java/org/folio/circulation/resources/ItemsByInstanceResource.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import org.folio.circulation.support.Clients;
+import org.folio.circulation.support.http.server.JsonHttpResponse;
+import org.folio.circulation.support.http.server.WebContext;
+import java.lang.invoke.MethodHandles;
+import java.util.List;
+
+public class ItemsByInstanceResource extends Resource {
+
+  private static final Logger log = LogManager.getLogger(MethodHandles.lookup().lookupClass());
+  public ItemsByInstanceResource(HttpClient client) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/resources/ItemsByInstanceResource.java:20]:
```suggestion
  private static final Logger log = LogManager.getLogger(MethodHandles.lookup().lookupClass());
  
  public ItemsByInstanceResource(HttpClient client) {
```

File: src/main/java/org/folio/circulation/resources/ItemsByInstanceResource.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  public void register(Router router) {
+    router.get("/circulation/items-by-instance")
+      .handler(this::getInstanceItems);
+  }
+
+  private void getInstanceItems(RoutingContext routingContext) {
+    final WebContext context = new WebContext(routingContext);
+    final Clients clients = Clients.create(context, client);
+    List<String> queryParams = routingContext.queryParam("query");
+    if (!queryParams.isEmpty()) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/resources/ItemsByInstanceResource.java:34]: And if `queryParams` is empty? Do we not send a response?

File: src/main/java/org/folio/circulation/resources/ItemsByInstanceResource.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+      final var searchRepository = new SearchRepository(clients);
+      searchRepository.getInstanceWithItems(queryParams.get(0))
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/resources/ItemsByInstanceResource.java:36]:
```suggestion
      new SearchRepository(clients).getInstanceWithItems(queryParams.get(0))
```

File: src/main/java/org/folio/circulation/resources/ItemsByInstanceResource.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        .thenApply(r -> r.map(this::toJson))
+        .thenApply(r -> r.map(JsonHttpResponse::ok))
+        .thenAccept(context::writeResultToHttpResponse);
+    }
+  }
+
+  private JsonObject toJson(SearchInstance instanceExtended) {
+    log.debug("toJson:: instanceExtended: {}", instanceExtended);
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/resources/ItemsByInstanceResource.java:44]:
```suggestion
    log.debug("toJson:: instanceExtended: {}", () -> instanceExtended);
```

File: src/main/java/org/folio/circulation/resources/ItemsByInstanceResource.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    if (instanceExtended != null)
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/resources/ItemsByInstanceResource.java:45]: I don't think that your instance can be null at this stage. And please don't use if-statements without curly braces.

File: src/test/java/api/ItemsByInstanceResourceTest.java [hunk 13]
```diff
@@ -0,0 +1,26 @@
+package api;
+
+import api.support.APITests;
+import org.folio.circulation.support.http.client.Response;
+import org.junit.jupiter.api.Test;
+import java.util.UUID;
+import static api.support.http.InterfaceUrls.itemsByInstanceUrl;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.Is.is;
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/ItemsByInstanceResourceTest.java:9]: wrong import order

File: src/main/java/org/folio/circulation/infrastructure/storage/SearchRepository.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+import org.folio.circulation.support.CollectionResourceClient;
+import org.folio.circulation.support.http.client.Response;
+import org.folio.circulation.support.results.Result;
+import java.lang.invoke.MethodHandles;
+import java.net.URLEncoder;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import static org.folio.circulation.support.results.Result.emptyAsync;
+import static org.folio.circulation.support.results.ResultBinding.flatMapResult;
+import static org.folio.circulation.support.results.ResultBinding.mapResult;
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/infrastructure/storage/SearchRepository.java:21]: wrong import order

File: src/test/java/api/support/builders/SearchInstanceBuilder.java [hunk 15]
```diff
@@ -0,0 +1,22 @@
+package api.support.builders;
+
+import io.vertx.core.json.JsonObject;
+import java.util.List;
+
+public class SearchInstanceBuilder extends JsonBuilder implements Builder {
+
+  private final JsonObject searchInstance;
+  public SearchInstanceBuilder(JsonObject searchInstance) {
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/support/builders/SearchInstanceBuilder.java:9]:
```suggestion
  private final JsonObject searchInstance;
  
  public SearchInstanceBuilder(JsonObject searchInstance) {
```

File: src/test/java/api/support/builders/SearchInstanceBuilder.java [hunk 16]
```diff
@@ -0,0 +1,22 @@
+package api.support.builders;
+
+import io.vertx.core.json.JsonObject;
+import java.util.List;
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/support/builders/SearchInstanceBuilder.java:4]: wrong import order

File: src/test/java/api/support/fakes/FakeOkapi.java [hunk 17]
```diff
@@ -414,6 +420,12 @@ public void start(Promise<Void> startFuture) throws IOException {
       .withChangeMetadata()
       .create().register(router);
 
+    new FakeStorageModuleBuilder()
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/support/fakes/FakeOkapi.java:423]: Why do you need 3 fake modules?

File: src/test/java/api/support/fakes/FakeSearchModule.java [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+import org.folio.circulation.support.http.server.WebContext;
+import org.folio.circulation.support.results.Result;
+import java.lang.invoke.MethodHandles;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import static java.lang.String.format;
+import static org.folio.circulation.support.results.CommonFailures.failedDueToServerError;
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/support/fakes/FakeSearchModule.java:22]: Wrong import order

File: src/test/java/api/support/fixtures/SearchInstanceFixture.java [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+
+  private final ResourceClient searchClient;
+
+  public SearchInstanceFixture() {
+    this.searchClient = ResourceClient.forSearchClient();
+  }
+
+  public void basedUponDunkirk(UUID instanceId) {
+    SearchInstanceBuilder builder = new SearchInstanceBuilder(new InstanceBuilder("Dunkirk",
+      UUID.randomUUID()).withId(instanceId).create()).withItems(List.of(ItemExamples.basedUponDunkirk(UUID.randomUUID(), UUID.randomUUID()).create()));
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/support/fixtures/SearchInstanceFixture.java:19]: please brake this multiple lines

File: src/main/java/org/folio/circulation/infrastructure/storage/SearchRepository.java [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+      return CompletableFuture.completedFuture(failed(new BadRequestFailure(
+        "Search result is empty")));
+    }
+    return AsyncCoordinationUtil.allOf(searchInstance.getItems(), this::fetchItemDetails)
+      .thenApply(r -> r.map(searchInstance::changeItems));
+  }
+
+  private CompletableFuture<Result<Item>> fetchItemDetails(Item searchItem) {
+    return itemRepository.fetchById(searchItem.getItemId())
+      .thenComposeAsync(itemRepository::fetchItemRelatedRecords)
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/infrastructure/storage/SearchRepository.java:63]: `itemRepository.fetchById()` already does this

**Review by OleksandrVidinieiev — APPROVED** [review]

**Pull request merged**